File: Leetcode/Python/_383.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        def to_dict(string):
            hash_map = {}
            for s in string:
                hash_map[s] = hash_map.get(s, 0) + 1
                
            return hash_map
        
        r = to_dict(ransomNote)
        m = to_dict(magazine)
        
        for c in ransomNote:
            if r[c] > m.get(c, 0):
                return False
            
        return True

File: Leetcode/cpp/383.cpp
class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        std::unordered_map<char, int> umapMag;

        for (auto chr : magazine) {
            ++umapMag[chr];
        }
        for (const auto& chr : ransomNote) {
            if (!umapMag.contains(chr) || umapMag[chr] == 0) return false;
            --umapMag[chr];
        }

        return true;
    }
};

class SolutionV2 {
public:
    bool canConstruct(string ransomNote, string magazine) {
        std::unordered_map<char, int> umapMag, umapRan;

        for (auto chr : magazine) {
            ++umapMag[chr];
        }
        for (auto chr : ransomNote) {
            ++umapRan[chr];
        }

        for (const auto& [key, value] : umapRan) {
            if (!umapMag.contains(key) || umapMag[key] < value) {
                return false;
            }
        }
        return true;
    }
};
